src/query/v2/shard_request_manager.hpp

```diff
@@ -407,6 +399,257 @@ class ShardRequestManager : public ShardRequestManagerInterface {
+        paginated_response_tracker.at(*shard_it) = PaginatedResponseState::PartiallyFinished;
+        ++shard_it;
+      }
+      responses.push_back(std::move(response));
+    }
+
+    return at_least_one_result;
+  }
+
+  bool PollOnRequests(ExecutionState<CreateVerticesRequest> &state, std::vector<CreateVerticesResponse> &responses) {
+    auto &shard_cache_ref = state.shard_cache;
+    size_t id = 0;
+    bool at_least_one_result = false;
+
+    for (auto shard_it = shard_cache_ref.begin(); shard_it != shard_cache_ref.end(); ++id) {
+      // This is fine because all new_vertices of each request end up on the same shard
+      const auto labels = state.requests[id].new_vertices[0].label_ids;
+
+      auto &storage_client = GetStorageClientForShard(*shard_it, labels[0].id);
+
+      auto poll_result = storage_client.PollAsyncWriteRequest();
+      if (!poll_result) {
+        continue;
+      }
+
+      at_least_one_result = true;
+
+      if (poll_result->HasError()) {
+        throw std::runtime_error("CreateVertices request timed out");
+      }
+
+      WriteResponses response_variant = poll_result->GetValue();
+      auto response = std::get<CreateVerticesResponse>(response_variant);
+
+      if (!response.success) {
+        throw std::runtime_error("CreateVertices request did not succeed");
+      }
+      responses.push_back(response);
+      shard_it = shard_cache_ref.erase(shard_it);
+    }
+
+    return at_least_one_result;
+  }
+
+  bool PollOnRequests(ExecutionState<ExpandOneRequest> &state, std::vector<ExpandOneResponse> &responses) {
+    auto &shard_cache_ref = state.shard_cache;
+    size_t id = 0;
+    bool at_least_one_result = false;
+
+    for (auto shard_it = shard_cache_ref.begin(); shard_it != shard_cache_ref.end(); ++id) {
+      auto &storage_client = GetStorageClientForShard(
+          *state.label, storage::conversions::ConvertPropertyVector(state.requests[id].src_vertices[0].second));
+
+      auto poll_result = storage_client.PollAsyncReadRequest();
+      if (!poll_result) {
+        continue;
+      }
+
+      at_least_one_result = true;
+
+      if (poll_result->HasError()) {
+        throw std::runtime_error("ExpandOne request timed out");
+      }
+
+      ReadResponses response_variant = poll_result->GetValue();
+      auto response = std::get<ExpandOneResponse>(response_variant);
+      // -NOTE-
+      // Currently a boolean flag for signaling the overall success of the
+      // ExpandOne request does not exist. But it should, so here we assume
+      // that it is allready in place.
+      if (!response.success) {
+        throw std::runtime_error("ExpandOne request did not succeed");
+      }
+
+      responses.push_back(std::move(response));
+      shard_it = shard_cache_ref.erase(shard_it);
+    }
+
+    return at_least_one_result;
+  }
+
+  void AwaitOnFirstRequest(ExecutionState<ScanVerticesRequest> &state, std::vector<ScanVerticesResponse> &responses,
```

Comment:
Let's rename this and slightly alter its behaviour:
1. Instead of iterating over the next pending request only. Block on them one by one. 
2. If a pending request is ready (that is, there was no Redirection from the RsmClient), update its status on the Paginated tracker. 
3. If there was a redirection, immediately resent the request on the RsmClient (see there the changes).